src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,683 @@
+
+	uint8_t msb = RegisterRead(Register::OTP_DATA_MSB);
+	uint8_t lsb = RegisterRead(Register::OTP_DATA_LSB);
+	*lsb_msb = ((msb << 8) | lsb) & 0xffff;
+	return 1;
+}
+
+void BMM350::UpdateMagOffsets()
+{
+	PX4_DEBUG("DUMPING OTP");
+	uint16_t otp_word = 0;
+	uint16_t otp_data[32] = {0};
+
+	for (int idx = 0; idx < 32; idx++) {
+		ReadOTPWord(idx, &otp_word);
+		otp_data[idx] = otp_word;
+		PX4_DEBUG("i: %i, val = %i", idx, otp_data[idx]);
+	}
+
+	PX4_DEBUG("var_id: %i", (otp_data[30] & 0x7f00) >> 9);
+
+	PX4_DEBUG("UPDATING OFFSETS");
+	uint16_t off_x_lsb_msb, off_y_lsb_msb, off_z_lsb_msb, t_off;
+	uint8_t sens_x, sens_y, sens_z, t_sens;
+	uint8_t tco_x, tco_y, tco_z;
+	uint8_t tcs_x, tcs_y, tcs_z;
+	uint8_t cross_x_y, cross_y_x, cross_z_x, cross_z_y;
+
+	off_x_lsb_msb = otp_data[0x0E] & 0x0FFF;
+	off_y_lsb_msb = ((otp_data[0x0E] & 0xF000) >> 4) +
+			(otp_data[0x0F] & 0x00FF);
+	off_z_lsb_msb = (otp_data[0x0F] & 0x0F00) +
+			(otp_data[0x10] & 0x00FF);
+	t_off = otp_data[0x0D] & 0x00FF;
+
+	_mag_comp_vals.dut_offset_coef.offset_x = FixSign(off_x_lsb_msb, 12);
+	_mag_comp_vals.dut_offset_coef.offset_y = FixSign(off_y_lsb_msb, 12);
+	_mag_comp_vals.dut_offset_coef.offset_z = FixSign(off_z_lsb_msb, 12);
+	_mag_comp_vals.dut_offset_coef.t_offs = FixSign(t_off, 8) / 5.0f;
+
+	sens_x = (otp_data[0x10] & 0xFF00) >> 8;
+	sens_y = (otp_data[0x11] & 0x00FF);
+	sens_z = (otp_data[0x11] & 0xFF00) >> 8;
+	t_sens = (otp_data[0x0D] & 0xFF00) >> 8;
+
+	_mag_comp_vals.dut_sensit_coef.sens_x = FixSign(sens_x, 8) / 256.0f;
+	_mag_comp_vals.dut_sensit_coef.sens_y = (FixSign(sens_y, 8) / 256.0f) + 0.01f;
+	_mag_comp_vals.dut_sensit_coef.sens_z = FixSign(sens_z, 8) / 256.0f;
+	_mag_comp_vals.dut_sensit_coef.t_sens = FixSign(t_sens, 8) / 512.0f;
+
+	tco_x = (otp_data[0x12] & 0x00FF);
+	tco_y = (otp_data[0x13] & 0x00FF);
+	tco_z = (otp_data[0x14] & 0x00FF);
+
+	_mag_comp_vals.dut_tco.tco_x = FixSign(tco_x, 8) / 32.0f;
+	_mag_comp_vals.dut_tco.tco_y = FixSign(tco_y, 8) / 32.0f;
+	_mag_comp_vals.dut_tco.tco_z = FixSign(tco_z, 8) / 32.0f;
+
+	tcs_x = (otp_data[0x12] & 0xFF00) >> 8;
+	tcs_y = (otp_data[0x13] & 0xFF00) >> 8;
+	tcs_z = (otp_data[0x14] & 0xFF00) >> 8;
+
+	_mag_comp_vals.dut_tcs.tcs_x = FixSign(tcs_x, 8) / 16384.0f;
+	_mag_comp_vals.dut_tcs.tcs_y = FixSign(tcs_y, 8) / 16384.0f;
+	_mag_comp_vals.dut_tcs.tcs_z = (FixSign(tcs_z, 8) / 16384.0f) - 0.0001f;
+
+	_mag_comp_vals.dut_t0 = (FixSign(otp_data[0x18], 16) / 512.0f) + 23.0f;
+
+	cross_x_y = (otp_data[0x15] & 0x00FF);
+	cross_y_x = (otp_data[0x15] & 0xFF00) >> 8;
+	cross_z_x = (otp_data[0x16] & 0x00FF);
+	cross_z_y = (otp_data[0x16] & 0xFF00) >> 8;
+
+	_mag_comp_vals.cross_axis.cross_x_y = FixSign(cross_x_y, 8) / 800.0f;
+	_mag_comp_vals.cross_axis.cross_y_x = FixSign(cross_y_x, 8) / 800.0f;
+	_mag_comp_vals.cross_axis.cross_z_x = FixSign(cross_z_x, 8) / 800.0f;
+	_mag_comp_vals.cross_axis.cross_z_y = FixSign(cross_z_y, 8) / 800.0f;
+}
+
+uint8_t BMM350::RegisterRead(Register reg)
```

Comment:
should return the read status value separate from register value. that is, better prototype would be
"int BMM350::RegisterRead(Register reg, uint8_t *val)"

This allows returning also register value "0xff", and checking whether the read succeeded. I admit that it would add some extra error checking code in the driver, but the case where i2c read would fail and the value 0xff is used would cause confusion.